onnx/inliner/inliner.cc

```diff
@@ -262,11 +310,120 @@ class Specializer {
+    result.reserve(node.input_size());
+    VisitNode(node);
+  }
+};
+
+std::vector<std::string> GetUsedVars(const NodeProto& node) {
+  return ComputeUsedVars(node).result;
+}
+
+using ConstNodeMap = std::unordered_map<std::string, const NodeProto*>;
+
+ConstNodeMap FindConstantNodes(const GraphProto& graph) {
+  ConstNodeMap result;
+  for (const NodeProto& node : graph.node()) {
+    if (IsOnnxDomain(node.domain()) && (node.op_type() == "Constant")) {
+      result[node.output(0)] = &node;
+    }
+  }
+  return result;
+}
+
+const TypeProto& GetType(const ModelProto& model, std::string var) {
+  for (auto& vi : model.graph().value_info()) {
+    if (vi.name() == var)
+      return vi.type();
+  }
+  for (auto& vi : model.graph().input()) {
+    if (vi.name() == var)
+      return vi.type();
+  }
+  for (auto& vi : model.graph().output()) {
+    if (vi.name() == var)
+      return vi.type();
+  }
+  ONNX_ASSERTM(false, "Type unknown for ", var);
+}
+
+void ConvertVersion(ModelProto& model, const NodeProto& call_node, FunctionProto& function, int target_version) {
+  shape_inference::InferShapes(model);
+
+  ModelProto function_as_model;
+  function_as_model.set_ir_version(model.ir_version());
+  *function_as_model.mutable_opset_import() = function.opset_import();
+
+  GraphProto& graph = *function_as_model.mutable_graph();
+  // The graph's inputs are all the variables used in the call_node.
+  auto used_vars = GetUsedVars(call_node);
+  auto constant_node_map = FindConstantNodes(model.graph());
+
+  RepeatedNodeProto& function_nodes = *function.mutable_node();
+  RepeatedNodeProto& nodes = *graph.mutable_node();
+  nodes.Reserve(function_nodes.size() + used_vars.size());
+
+  auto* inputs = graph.mutable_input();
+  for (const auto& var : used_vars) {
+    auto* new_input = inputs->Add();
+    new_input->set_name(var);
+    *new_input->mutable_type() = GetType(model, var);
+    // Create a copy of constants used by the call_node.
+    // We do not handle initializers-as-constants for now.
+    auto it = constant_node_map.find(var);
+    if (it != constant_node_map.end()) {
+      *nodes.Add() = *(it->second);
+    }
+  }
+
+  // outputs: from call_node node outputs
+  auto* outputs = graph.mutable_output();
+  for (const auto& var : call_node.output()) {
+    if (!var.empty()) {
+      auto* new_output = outputs->Add();
+      new_output->set_name(var);
+      *new_output->mutable_type() = GetType(model, var);
+    }
+  }
+
+  // TODO: Use std::move when it is fully supported on all protobuf platforms used
+  for (auto& function_node : function_nodes)
+    *nodes.Add() = function_node;
+  function_nodes.Clear();
+
+  auto converted = ONNX_NAMESPACE::version_conversion::ConvertVersion(function_as_model, target_version);
+
+  function_nodes.Swap(converted.mutable_graph()->mutable_node());
+
+  // Append new initializers to main graph initializers
+  auto& added_initializer = converted.graph().initializer();
```

Comment:
sparse initializers?

Reply:
Added. However, none of existing version-converters add a sparse initializer. Some of them do add a new initializer. I think it is a better idea to eliminate even that, and uniformly add a new Constant node (instead of an initializer), which some of the version-converter rules do. (In the long run, ideally we would not require either of these.)